homeassistant/components/sensor/__init__.py

```diff
@@ -410,24 +485,43 @@ def state(self) -> Any:
 
         # Sensors with device classes indicating a non-numeric value
         # should not have a state class or unit of measurement
-        if device_class in {
-            SensorDeviceClass.DATE,
-            SensorDeviceClass.ENUM,
-            SensorDeviceClass.TIMESTAMP,
-        }:
-            if self.state_class:
-                raise ValueError(
-                    f"Sensor {self.entity_id} has a state class and thus indicating "
-                    "it has a numeric value; however, it has the non-numeric "
-                    f"device class: {device_class}"
-                )
+        if (
+            device_class
+            in {
+                SensorDeviceClass.DATE,
+                SensorDeviceClass.ENUM,
+                SensorDeviceClass.TIMESTAMP,
+            }
+            and unit_of_measurement
+        ):
+            raise ValueError(
+                f"Sensor {self.entity_id} has a unit of measurement and thus "
+                "indicating it has a numeric value; however, it has the "
+                f"non-numeric device class: {device_class}"
+            )
 
-            if unit_of_measurement:
-                raise ValueError(
-                    f"Sensor {self.entity_id} has a unit of measurement and thus "
-                    "indicating it has a numeric value; however, it has the "
-                    f"non-numeric device class: {device_class}"
-                )
+        # Validate state class for sensors with a device class
+        if (
+            not self._invalid_state_class_reported
+            and device_class
+            and (classes := DEVICE_CLASS_STATE_CLASSES.get(device_class)) is not None
+            and state_class not in classes
+        ):
```

Comment:
If `None` is always allowed, then it can be simplified?
Or is the objective that some device classes disallow `None`?
```suggestion
        if (
            state_class
            and not self._invalid_state_class_reported
            and device_class
            and (classes := DEVICE_CLASS_STATE_CLASSES.get(device_class)) is not None
            and state_class not in classes
        ):
```

Reply:
Hmm, yeah that was my initial driver, but that is not now. I'll clean up.